### PR TITLE
feat(embodichain): integrate cartpole into rlinf

### PR DESCRIPTION
**Pull request by yangchen73**
This PR integrates EmbodiChain (https://github.com/DexForce/EmbodiChain) CartPole into RLinf's embodied training pipeline.

### Description

- add a new `embodichain` environment type and register it in RLinf
- implement `EmbodiChainEnv` as an adapter between RLinf and EmbodiChain's RL task interface
- add RLinf training configs for EmbodiChain CartPole
- fix a bug about ensuring chunked tnesors are contiguous on split chunks
- Update docs and readme

### Motivation and Context
- RLinf currently does not support EmbodiChain environments.
- This PR adds a first working integration for EmbodiChain CartPole, allowing RLinf to train and evaluate policies on this rl task.

### How has this been tested?
Tested by installing RLinf embodied dependencies together with EmbodiChain:

```bash
cd <path_to_RLinf_repository>
bash requirements/install.sh embodied --env embodichain
```
Then verified the EmbodiChain CartPole example by launching training with the reference config.

Since the CartPole gym config is not yet included in the current EmbodiChain release (We plan to include it in the next version), testing currently requires a local EmbodiChain responsitry with config in it:

```bash
git clone https://github.com/DexForce/EmbodiChain.git /path/to/EmbodiChain
export EMBODICHAIN_PATH=/path/to/EmbodiChain
bash examples/embodiment/run_embodiment.sh embodichain_ppo_cart_pole
```

### Additional information (optional, e.g., figures and logs):
- Cartpole has been shown to converge.
截图 2026-03-26 17-25-49

- This PR also adds demo GIF assets and updates the Embodied Examples index pages in both English and Chinese.
- The current documentation is an initial draft and will be expanded to cover more EmbodiChain environments in future updates.
 
### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update (Document-only update)

### Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by yuecideng** [comment]
Hi RLinf Team, We are core contributors to EmbodiChain, and this PR is motivated by a shared vision with RLinf, building a scalable and unified infrastructure for embodied intelligence and agentic learning.

## Motivation
EmbodiChain focuses on enabling end-to-end research and development of embodied intelligence and bridging the gap from simulator to achieve seamless sim2real transfer to the real world. RLinf provides a highly flexible and efficient reinforcement learning infrastructure for training large-scale models across embodied and agent domains. By integrating EmbodiChain environments (starting with CartPole) into RLinf, we aim to bridge simulation and data generation abstraction with training infrastructure, lowering the barrier for rapid experimentation and system-level innovation in embodied RL.

## Collaboration Intent
This integration is only a first step. We see strong potential for deeper collaboration between EmbodiChain and RLinf in several directions:
- Standardizing environment interfaces for embodied and agentic RL
- Enabling seamless coupling between structured interaction pipelines and RL training loops
- Co-designing benchmarks that better reflect real-world embodied intelligence challenges

We believe that aligning on system abstractions and interfaces can significantly accelerate the development of next-generation embodied AI systems.

## Future Plans.
Looking ahead, we plan to:
- Extend EmbodiChain integration to more complex embodied environments and task.
- Explore integration with world models, VLA policies, and maybe new training paradigm.
- Contribute to a more unified ecosystem for embodied + agentic learning

We are excited about the opportunity to collaborate with the RLinf community and welcome further discussions, feedback, and joint development efforts.

File: rlinf/envs/embodichain/embodichain_env.py [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+            if "episode" in infos:
+                if "success" in infos:
+                    infos["episode"]["success_at_end"] = infos["success"].clone()
+                if "fail" in infos:
+                    infos["episode"]["fail_at_end"] = infos["fail"].clone()
+        dones = torch.logical_or(terminations, truncations)
+        if dones.any() and self.auto_reset:
+            done_ids = torch.nonzero(dones, as_tuple=False).flatten().to(torch.int32)
+            self._reset_metrics(done_ids)
+        self._is_start = False
```
[the remark this answers is not in the document]
Comment (yangchen73) [reply rlinf/envs/embodichain/embodichain_env.py:378]: In our setup `EnvWorker` only calls `chunk_step()`, not `step()` alone. `chunk_step()` already adds `final_info`  `final_observation` on the last chunk step when episodes end, which is what the worker needs. `step()` stays simple because it is only used inside `chunk_step()`.

**Comment by Iron-Wph** [comment]
@yangchen73 @yuecideng 
Hello, Thanks for your contribution!

I noticed that the current PR title does not conform to the Conventional Commits specification required by the project. Could you please update it to follow the expected format?

For example:
feat(embodichain): integrate cartpole into rlinf

I followed the installation instructions in the documentation,
```
bash requirements/install.sh embodied --env embodichain
```
but encountered the error shown below. 
PixPin_2026-03-28_19-59-30

It seems to be caused by a dependency issue in EmbodiChain.

**Comment by yangchen73** [comment]
> @yangchen73 @yuecideng Hello, Thanks for your contribution!
> 
> I noticed that the current PR title does not conform to the Conventional Commits specification required by the project. Could you please update it to follow the expected format?
> 
> For example: feat(embodichain): integrate cartpole into rlinf
> 
> I followed the installation instructions in the documentation,
> 
> ```
> bash requirements/install.sh embodied --env embodichain
> ```
> 
> but encountered the error shown below. PixPin_2026-03-28_19-59-30
> 
> It seems to be caused by a dependency issue in EmbodiChain.

Hi, thanks for the feedback!

I will update the PR title to follow the Conventional Commits specification immediately.

Regarding the installation error, it is indeed a dependency issue. You can try to resolve it temporarily by running:
```
uv pip install embodichain --extra-index-url http://pyp.open3dv.site:2345/simple/ --trusted-host pyp.open3dv.site
```
Please let me know if this works for you!

File: requirements/install.sh [hunk 2]
```diff
@@ -844,6 +848,13 @@ install_frankasim_env() {
     uv pip install -r "$serldir/franka_sim/requirements.txt"
 }
 
+install_embodichain_env() {
+    local embodichain_dir
+    embodichain_dir=$(clone_or_reuse_repo EMBODICHAIN_PATH "$VENV_DIR/EmbodiChain" https://github.com/DexForce/EmbodiChain.git)
+    uv pip install -e "$embodichain_dir"
```
Comment (yuecideng) [inline requirements/install.sh]: We should add extra index `--extra-index-url http://pyp.open3dv.site:2345/simple/ --trusted-host pyp.open3dv.site`  @yangchen73 .

**Comment by Iron-Wph** [comment]
> > @yangchen73 @yuecideng Hello, Thanks for your contribution!
> > I noticed that the current PR title does not conform to the Conventional Commits specification required by the project. Could you please update it to follow the expected format?
> > For example: feat(embodichain): integrate cartpole into rlinf
> > I followed the installation instructions in the documentation,
> > ```
> > bash requirements/install.sh embodied --env embodichain
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but encountered the error shown below. PixPin_2026-03-28_19-59-30
> > It seems to be caused by a dependency issue in EmbodiChain.
> 
> Hi, thanks for the feedback!
> 
> I will update the PR title to follow the Conventional Commits specification immediately.
> 
> Regarding the installation error, it is indeed a dependency issue. You can try to resolve it temporarily by running:
> 
> ```
> uv pip install embodichain --extra-index-url http://pyp.open3dv.site:2345/simple/ --trusted-host pyp.open3dv.site
> ```
> 
> Please let me know if this works for you!

@yangchen73 
Thanks for your reply! It would be helpful to include this step in the requirements/install.sh script.

I also encountered another issue related to `embodichain_data` when running:
```
bash examples/embodiment/run_embodiment.sh embodichain_ppo_cart_pole
```
Currently, the program tries to download the required resources at runtime, which may fail (e.g., due to network or mirror issues). In addition, the downloaded file sometimes fails the MD5 check and keeps retrying.

More details are as follow.
1. log files: [run_embodiment.log](https://github.com/user-attachments/files/26328694/run_embodiment.log)

2. example pic.
PixPin_2026-03-29_14-01-03


Could you please provide a way to:
1. Pre-download the required dataset/resources manually (e.g., an official download link), and  
2. Specify the dataset path via an environment variable or configuration,

so that we can run the program in an offline or controlled environment without triggering repeated downloads?

Thanks in advance!

File: requirements/install.sh (hunk 2 above)
Comment (yangchen73) [reply requirements/install.sh]: Done. I'v added this extra index.

**Comment by Iron-Wph** [comment]
> > > @yangchen73 @yuecideng Hello, Thanks for your contribution!
> > > I noticed that the current PR title does not conform to the Conventional Commits specification required by the project. Could you please update it to follow the expected format?
> > > For example: feat(embodichain): integrate cartpole into rlinf
> > > I followed the installation instructions in the documentation,
> > > ```
> > > bash requirements/install.sh embodied --env embodichain
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > but encountered the error shown below. PixPin_2026-03-28_19-59-30
> > > It seems to be caused by a dependency issue in EmbodiChain.
> > 
> > 
> > Hi, thanks for the feedback!
> > I will update the PR title to follow the Conventional Commits specification immediately.
> > Regarding the installation error, it is indeed a dependency issue. You can try to resolve it temporarily by running:
> > ```
> > uv pip install embodichain --extra-index-url http://pyp.open3dv.site:2345/simple/ --trusted-host pyp.open3dv.site
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Please let me know if this works for you!
> 
> @yangchen73 Thanks for your reply! It would be helpful to include this step in the requirements/install.sh script.
> 
> I also encountered another issue related to `embodichain_data` when running:
> 
> ```
> bash examples/embodiment/run_embodiment.sh embodichain_ppo_cart_pole
> ```
> 
> Currently, the program tries to download the required resources at runtime, which may fail (e.g., due to network or mirror issues). In addition, the downloaded file sometimes fails the MD5 check and keeps retrying.
> 
> More details are as follow.
> 
> 1. log files: [run_embodiment.log](https://github.com/user-attachments/files/26328694/run_embodiment.log)
> 2. example pic.
> 
> PixPin_2026-03-29_14-01-03
> Could you please provide a way to:
> 
> 1. Pre-download the required dataset/resources manually (e.g., an official download link), and
> 2. Specify the dataset path via an environment variable or configuration,
> 
> so that we can run the program in an offline or controlled environment without triggering repeated downloads?
> 
> Thanks in advance!

@yangchen73 
Hello, I’m still encountering issues with the dataset.
I attempted to download the files from Hugging Face (https://huggingface.co/datasets/dexforce/embodichain_data) , but the process fails due to an MD5 checksum mismatch, which interrupts the program.

Additionally, I reviewed the EmbodiChain installation documentation (https://dexforce.github.io/EmbodiChain/quick_start/install.html), but I couldn’t find any guidance on how to properly download or prepare the dataset.

So I was wondering if there is a specific solution to this issue, or whether it’s possible to bypass the problem by pre-downloading the dataset and letting the program use the local files instead.

**Comment by yangchen73** [comment]
> > > > @yangchen73 @yuecideng Hello, Thanks for your contribution!
> > > > I noticed that the current PR title does not conform to the Conventional Commits specification required by the project. Could you please update it to follow the expected format?
> > > > For example: feat(embodichain): integrate cartpole into rlinf
> > > > I followed the installation instructions in the documentation,
> > > > ```
> > > > bash requirements/install.sh embodied --env embodichain
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > but encountered the error shown below. PixPin_2026-03-28_19-59-30
> > > > It seems to be caused by a dependency issue in EmbodiChain.
> > > 
> > > 
> > > Hi, thanks for the feedback!
> > > I will update the PR title to follow the Conventional Commits specification immediately.
> > > Regarding the installation error, it is indeed a dependency issue. You can try to resolve it temporarily by running:
> > > ```
> > > uv pip install embodichain --extra-index-url http://pyp.open3dv.site:2345/simple/ --trusted-host pyp.open3dv.site
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Please let me know if this works for you!
> > 
> > 
> > @yangchen73 Thanks for your reply! It would be helpful to include this step in the requirements/install.sh script.
> > I also encountered another issue related to `embodichain_data` when running:
> > ```
> > bash examples/embodiment/run_embodiment.sh embodichain_ppo_cart_pole
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Currently, the program tries to download the required resources at runtime, which may fail (e.g., due to network or mirror issues). In addition, the downloaded file sometimes fails the MD5 check and keeps retrying.
> > More details are as follow.
> > 
> > 1. log files: [run_embodiment.log](https://github.com/user-attachments/files/26328694/run_embodiment.log)
> > 2. example pic.
> > 
> > PixPin_2026-03-29_14-01-03
> > Could you please provide a way to:
> > 
> > 1. Pre-download the required dataset/resources manually (e.g., an official download link), and
> > 2. Specify the dataset path via an environment variable or configuration,
> > 
> > so that we can run the program in an offline or controlled environment without triggering repeated downloads?
> > Thanks in advance!
> 
> @yangchen73 Hello, I’m still encountering issues with the dataset. I attempted to download the files from Hugging Face (https://huggingface.co/datasets/dexforce/embodichain_data) , but the process fails due to an MD5 checksum mismatch, which interrupts the program.
> 
> Additionally, I reviewed the EmbodiChain installation documentation (https://dexforce.github.io/EmbodiChain/quick_start/install.html), but I couldn’t find any guidance on how to properly download or prepare the dataset.
> 
> So I was wondering if there is a specific solution to this issue, or whether it’s possible to bypass the problem by pre-downloading the dataset and letting the program use the local files instead.

Thanks for following up and for providing such detailed feedback! We appreciate you taking the time to test the Hugging Face download as well.

We are currently discussing this internally and looking into the best way to implement it. We'll keep you updated here once we figure out a solid solution or have a patch ready.

**Comment by yuecideng** [comment]
Hi @Iron-Wph , we just update the code and you can try with the following steps:
1. update the embodichain code base.
2. set env variable for testing `export EMBODICHAIN_DATA_ROOT=/path/to/data`
3. run `python -m embodichain.data download --name CartPole` and `python -m embodichain.data download --name SimResources`
4. If the download is success, run example training as mentioned above.

**Comment by Iron-Wph** [comment]
> Hi @Iron-Wph , we just update the code and you can try with the following steps:
> 
> 1. update the embodichain code base.
> 2. set env variable for testing `export EMBODICHAIN_DATA_ROOT=/path/to/data`
> 3. run `python -m embodichain.data download --name CartPole` and `python -m embodichain.data download --name SimResources`
> 4. If the download is success, run example training as mentioned above.

@yangchen73 
Hello, thank you for providing the dataset download instructions — they work very well.

However, during training, I repeatedly encounter the same error, which seems to be caused by inconsistent data types returned at a lower level. Could you please help double-check this?
The error message is shown in the figure below. 
PixPin_2026-03-31_20-01-18

For more detailed information, please refer to the following two files.
1) log file:  [run_embodiment.log](https://github.com/user-attachments/files/26377411/run_embodiment.log)

2) config file:  [config.yaml](https://github.com/user-attachments/files/26377420/config.yaml)


Additionally, when setting up the environment following the provided instructions, I ran into an issue where importing the dexsim module fails due to a missing dependency: libpython3.11.so.1.0. It would be helpful if the installation of this dependency could be included in the setup script. 
The error message is shown in the figure below. 
PixPin_2026-03-31_20-02-18

**Comment by yuecideng** [comment]
Hi @Iron-Wph , thanks for your suggestion. We haven't adopted the 'uv' package manager yet, but we plan to address the dependency issues in an upcoming release.

**Comment by yangchen73** [comment]
> > Hi @Iron-Wph , we just update the code and you can try with the following steps:
> > 
> > 1. update the embodichain code base.
> > 2. set env variable for testing `export EMBODICHAIN_DATA_ROOT=/path/to/data`
> > 3. run `python -m embodichain.data download --name CartPole` and `python -m embodichain.data download --name SimResources`
> > 4. If the download is success, run example training as mentioned above.
> 
> @yangchen73 Hello, thank you for providing the dataset download instructions — they work very well.
> 
> However, during training, I repeatedly encounter the same error, which seems to be caused by inconsistent data types returned at a lower level. Could you please help double-check this? The error message is shown in the figure below. PixPin_2026-03-31_20-01-18
> 
> For more detailed information, please refer to the following two files.
> 
> 1. log file:  [run_embodiment.log](https://github.com/user-attachments/files/26377411/run_embodiment.log)
> 2. config file:  [config.yaml](https://github.com/user-attachments/files/26377420/config.yaml)
> 
> Additionally, when setting up the environment following the provided instructions, I ran into an issue where importing the dexsim module fails due to a missing dependency: libpython3.11.so.1.0. It would be helpful if the installation of this dependency could be included in the setup script. The error message is shown in the figure below. PixPin_2026-03-31_20-02-18

Hi @Iron-Wph , 
Thanks for the logs and config. They helped a lot.

The crash happens because with multi-GPU, some env ranks may have no finished episodes in one rollout, so their metrics can be empty or show up as [ ] on the driver. The old code always called torch.concat on tensors only, so it broke when [ ] appeared.

We fixed this in RLinf by normalizing each rank’s values and merging all metric keys across ranks before concat. Pull the latest on this branch and your training should pass that step.

截图 2026-04-01 11-06-44

**Comment by Iron-Wph** [comment]
@yangchen73 
Thanks for your help ! 
The results I reproduced are shown below and appear to be correct.
PixPin_2026-04-01_14-17-57

However, this PR has a commit check issue, as shown below.
PixPin_2026-04-01_14-22-30

**Comment by yangchen73** [comment]
> @yangchen73 Thanks for your help ! The results I reproduced are shown below and appear to be correct. PixPin_2026-04-01_14-17-57
> 
> However, this PR has a commit check issue, as shown below. PixPin_2026-04-01_14-22-30

I'll close this PR and open a new PR with correct branch name.
